Dockerfile

```diff
@@ -1,23 +1,35 @@
-FROM golang:1.9.0-alpine3.6 AS build
+# GitHub:  https://github.com/gohugoio/
+# Docker:  https://hub.docker.com/r/gohugoio/
+# Twitter: https://twitter.com/gohugoio
 
-RUN apk add --no-cache --virtual git musl-dev
-RUN go get github.com/golang/dep/cmd/dep
+FROM golang:1.10-rc-alpine3.7 AS build
 
-WORKDIR /go/src/github.com/gohugoio/hugo
-RUN dep ensure
-ADD . /go/src/github.com/gohugoio/hugo/
-RUN go install -ldflags '-s -w'
-
-FROM alpine:3.6
 RUN \
-  adduser -h /site -s /sbin/nologin -u 1000 -D hugo && \
   apk add --no-cache \
-    dumb-init
+    git \
+    musl-dev && \
+  go get github.com/golang/dep/cmd/dep && \
+  go get github.com/kardianos/govendor && \
+  govendor get github.com/gohugoio/hugo && \
+  cd /go/src/github.com/gohugoio/hugo && \
+  dep ensure && \
```

Comment:
Although you consider it breaks docker best practices, you may consider that doing this requires re-building
and re-downloading everything for every single `docker build`.

It is okay when the build step is built by a remote server that always discards cache, but might be interesting when iterating on `docker build` locally. Docker has even edited its [best practices](https://github.com/docker/docker.github.io/blob/master/engine/userguide/eng-image/dockerfile_best-practices.md#minimize-the-number-of-layers) explaining latest versions of docker have "have mitigated the need" of minimizing the number of layers.

As suggested in #4077 distinguishing installing dependent tools, dependencies and actual binary generation allows to benefit the docker build cache while building the image.
Having a dockerfile of (as set in #4077 ):

```
RUN apk add --no-cache \
    git \
    musl-dev \
 && go get github.com/golang/dep/cmd/dep

COPY Gopkg.lock Gopkg.toml /go/src/github.com/gohugoio/hugo/
WORKDIR /go/src/github.com/gohugoio/hugo
RUN dep ensure -vendor-only
COPY . /go/src/github.com/gohugoio/hugo/
RUN go install -ldflags '-s -w'
```

considers that:
- `dep`, `git`, `musl-dev` are external tools that change the less, without a strong dependency on the exact version
- `go` dependencies are changing slower than the code itself, and quite [long to install](https://github.com/golang/dep/blob/master/docs/FAQ.md#why-is-dep-slow) that are worth caching in local developments, and needs to be re-built every time `Gopkg.*` changes
- `hugo` code changes the most often and thus needs to be re-build every time a file is changed